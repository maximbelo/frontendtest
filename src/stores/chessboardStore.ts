import { defineStore } from "pinia";
import type { SelectedSquare, HistoryItem } from "@/interfaces";

export const useChessboardStore = defineStore("chessboard", {
  state: () => ({
    // Initialize the selectedSquare state with null values for file and rank
    selectedSquare: {
      file: null,
      rank: null,
    } as SelectedSquare,
    // Initialize the history state as an empty array
    history: [] as HistoryItem[],
  }),

  actions: {
    // This is responsible for updating the selected square and the history
    selectSquare(file: string, rank: string): void {
      this.selectedSquare = { file, rank };
      this.history.push({ file, rank });

      console.log(`Selected square: ${file}${rank}`);
      console.log(this.history);
    },
    // Reset the states
    reset(): void {
      this.selectedSquare = { file: null, rank: null };
      this.history = [];
    },
  },
});
