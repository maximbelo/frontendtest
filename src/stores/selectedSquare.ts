import { defineStore } from "pinia";
import type { SelectedSquare, HistoryItem } from "@/types";

export const useSelectedSquareStore = defineStore("selectedSquare", {
  state: () => ({
    selectedSquare: {
      file: null,
      rank: null,
    } as SelectedSquare,
    history: [] as HistoryItem[],
  }),
  actions: {
    selectSquare(file: string, rank: string): void {
      this.selectedSquare = { file, rank };
      this.history.push({ file, rank });
      console.log(`Selected square: ${file}${rank}`);
      console.log(this.history);
    },
  },
});
