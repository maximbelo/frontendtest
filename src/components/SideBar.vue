<template>
  <aside class="sidebar">
    <h2>Selected square:</h2>
    <p>{{ store.selectedSquare.file }}{{ store.selectedSquare.rank }}</p>

    <h2>Selected history:</h2>
    <ul>
      <li v-for="item in store.history" :key="item.file + item.rank">
        {{ item.file }}{{ item.rank }}
      </li>
    </ul>
  </aside>
</template>

<script setup lang="ts">
import { useSelectedSquareStore } from "@/stores/selectedSquare";

const store = useSelectedSquareStore();
</script>

<style lang="scss" scoped></style>
