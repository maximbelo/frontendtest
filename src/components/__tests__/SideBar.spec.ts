import { describe, expect, test, beforeEach } from "vitest";
import { shallowMount } from "@vue/test-utils";
import { createPinia, setActivePinia } from "pinia";
import SideBar from "@/components/SideBar.vue";
import { useChessboardStore } from "@/stores/chessboardStore";

describe("SideBar.vue tests", () => {
  beforeEach(() => {
    setActivePinia(createPinia());
  });

  test("displays the selected square", () => {
    const store = useChessboardStore();
    store.selectedSquare = { file: "a", rank: "1" };

    const wrapper = shallowMount(SideBar);

    const selectedSquare = wrapper.find(".sidebar p");
    expect(selectedSquare.text()).toContain("a1");
  });

  test("displays the selected history", () => {
    const store = useChessboardStore();
    store.history = [
      { file: "a", rank: "1" },
      { file: "b", rank: "2" },
      { file: "c", rank: "3" },
    ];

    const wrapper = shallowMount(SideBar);

    const historyItems = wrapper.findAll(".sidebar li");
    expect(historyItems).toHaveLength(3);

    expect(historyItems[0].text()).toContain("1. a1");
    expect(historyItems[1].text()).toContain("2. b2");
    expect(historyItems[2].text()).toContain("3. c3");
  });

  test("matches the snapshot", () => {
    const store = useChessboardStore();
    store.selectedSquare = { file: "a", rank: "1" };
    store.history = [
      { file: "a", rank: "1" },
      { file: "b", rank: "2" },
      { file: "c", rank: "3" },
    ];

    const wrapper = shallowMount(SideBar);

    expect(wrapper.html()).toMatchSnapshot();
  });
});
