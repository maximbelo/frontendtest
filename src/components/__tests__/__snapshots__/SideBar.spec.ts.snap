// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SideBar.vue tests > matches the snapshot 1`] = `
"<aside class=\\"sidebar\\" data-v-ce868814=\\"\\"><button class=\\"reset-button\\" data-v-ce868814=\\"\\">Reset</button>
  <h2 data-v-ce868814=\\"\\">Selected square:</h2>
  <p data-v-ce868814=\\"\\">a1</p>
  <h2 data-v-ce868814=\\"\\">Selected history:</h2>
  <ul data-v-ce868814=\\"\\">
    <li data-v-ce868814=\\"\\">1. a1</li>
    <li data-v-ce868814=\\"\\">2. b2</li>
    <li data-v-ce868814=\\"\\">3. c3</li>
  </ul>
</aside>
<!-- ./sidebar -->"
`;
